tour.md

```diff
@@ -1,730 +1,493 @@
+
+```rust
+// v : null;
+v := null;
+
+// v : int;
+v := 0;
+// v : float;
+v := 0.0;
+
+// v : bool;
+v := true;
+
+// v : string
+v := "test";
+v := "formatted {v}"; // string interpolation
+v := "escaped curlies \{v}"; // exact value is `escaped curlies {v}`
+// strings also support other escaped charaters such as `\n`, `\t`, `\x2800`, etc.
+v := "\x2800";
+
+// v : (T,)
+v := (v,);
+// v : (T, T)
+v := (v, v);
+
+// v : [T]
+v := [];
+v := [v, v, v, v];
+v := [v; 1024];
+```
+
+Variable declaration
+
+```rust
+// inferred type
+name := 0;
+
+// explicit type
+// this declaration shadows the previous one, so its type may differ
+name: string = "value";
+```
+
+Operators
+
+- Arithmetic (`+`, `-`, `/`, `*`, `**`)
+- Equality (`==`, `!=`)
+- Comparison (`>`, `>=`, `<`, `<=`)
+- Boolean (`not`/`!`, `and`/`&&`, `or`/`||`)
+
+```rust
+2 + 2;
+2 - 2;
+2 / 2;
+2 * 2;
+2 % 2;
+2 ** 2;
+2 == 2;
+2 != 2;
+2 > 2;
+2 >= 2;
+2 < 2;
+2 <= 2;
+-2
+!true;
+true && true;
+false || true;
+a ?? b;
+```
+
+Assignment and compound assignments
+
+```rust
+name = 1;
+// compound assignments for all binary arithmetic and bitwise operators, and optional coalescing
+// `ident OP= expr desugar to `ident = ident OP expr`
+name += 1;
+name -= 1;
+name /= 1;
+name *= 1;
+name %= 1;
+name **= 1;
+// assigns to `name` only if `name` is null,
+// and asserts that `name` is no longer null
+name ??= 1;
+```
+
+If, do
+
+```rust
+if a() { /*...*/ }
+else if b() { /*...*/ }
+else { /*...*/ }
+
+do { /*...*/ }
+
+// both of the above are expressions:
+// the last expression in the body is the result
+v0 := if v { "a" } else { "c" }
+v1 := do { "value" }
+// to ensure that nothing is returned, put a semicolon after the last expression
+v2: null = do { "value"; }
+
+// you can use control statements without braces in `if` statements
+if true return
+if true break
+if true continue
+if true for ...
+if true while ...
+if true loop ...
+```
+
+Loops (for, while, loop), continue/break
+
+```rust
+for item in iterator { /*...*/ }
+
+while condition() { /*...*/ }
+loop { /*...*/ }
+
+for i in 0..10 {
+  print(i)
+}
+
+v := 10
+while v >= 0 {
+  print(v)
+  v -= 1
+}
+
+v := 0
+loop {
+  if v == 5 break
+  v += 1
+}
+```
+
+Functions (generics, variadics, return)
+
+```rust
+fn name(a: A, b: B, c: C) {/*...*/}
+fn name[T](a: A, b: B, c: C) -> T {/*...*/}
+fn name[T, E](a: A, b: B, c: C) -> T where T: Bound {/*...*/}
+
+fn name[T, E](a: A, b: B, c: C) -> T
+  where
+    A: IA,
+    B: IB,
+    C: IC
+{
+  /*...*/
+}
+
+// anonymous function
+square := \x {x*x}
+
+fn fib(n: int): int {
+  if n < 2 { n }
+  else n * fib(n - 1)
+}
+
+// return may be used to return early
+fn test() {
+  if something() { return }
+  print("yo")
+}
+
+fn split(s: string, sep: string) -> [string] {
+  out := [];
+  start := 0;
+  end := 0;
+
+  for ch in s {
+    if ch == sep {
+      out.push(s[start..end])
+      start = end;
+    } else {
+      end += 1;
+    }
+  }
+
+  if current.len > 0 {

```

Comment:
`current` is no longer :O. I think the check here should be `if (end-start > 0)`